### DISCUSSION TITLE
Feedback on the alpha builds

### DISCUSSION THREAD

**Discussion by christiaangoossens** [christiaangoossens/hass-oidc-auth#10] [General]
Do you have any feedback on the alpha builds? Did you get it running? Which parameters did you use?

**Comment by moutasem1989** [comment]
Today I installed hass-oidc-auth and configured it with Authentik. It worked as expected with no errors!

**Comment by FlyingNimbusCloud** [comment]
```
auth_oidc:
  client_id: "homeassistant"
  discovery_url: "https://auth.mydomain.net/.well-known/openid-configuration"
  features:
    automatic_user_linking: true
```

worked right away with my Authelia OIDC auth provider as a public client. I was bit confused by the "Just Checking" label on the input box but eventually figured out that's for my OTP.

It worked both by clicking the button for the browser session and it also worked to authenticate the app using the OTP

**Reply by christiaangoossens** [reply]
Good to hear that it works with Authelia. 

You also ran into the main downside at the moment: I have not found any options yet on how to make the Home Assistant UI look better (frontend is hard coded for the normal options, so I cannot send any new texts).

That's why I stuck with reusing the existing MFA UI with 'Just checking' and empty description and error messages, to differentiate it from username/password.

**Comment by Node815** [comment]
I'm testing this with[ pocket-ID](https://github.com/stonith404/pocket-id) and have encountered some issues and not sure if it's coming from the settings in the OIDC App or Home Assistant is sending it.  From Pocket ID's docker log:

```
[GIN] 2025/01/01 - 06:24:04 | 400 |     538.557µs |    XX.53.XX.26 | POST     "/api/oidc/authorize/new-client"
Error #01: invalid callback URL, it might be necessary for an admin to fix this
```

The question I have is the "api/oidc/authorize/new-client"   I'm not sure if this would be what Home Assistant is telling it or not.  In my OIDC Provider's settings, I've tried using both the /auth/oidc/welcome and /auth/oidc/redirect   url after my external domain with no success.  It sends correctly to the OIDC app but stops there.  

Once I get this fixed, I can write up a guide to setting it up with that program.  It actually replaced my original Authentik instance!  So, it's my main system.

**Reply by christiaangoossens** [reply]
Callback URL should be your external URL + `/auth/oidc/callback`. Let me know if that works for you.

**Reply by Node815** [reply]
Thanks!  That fixed the callback url problem, but now Home Assistant complained about an invalid token:
![image](https://github.com/user-attachments/assets/636ac230-7ec0-4693-a169-0525a399e89e)

So, I added that in which then sent gave me the next screen to auth the login as expected and threw:

![image](https://github.com/user-attachments/assets/95fe90d9-759f-41aa-8710-86ad802eddba)


![image](https://github.com/user-attachments/assets/0e926f2a-e8d1-4916-bffe-1c4a2eeb9684)

My config file:
![image](https://github.com/user-attachments/assets/df0f11ee-82d8-488c-807c-bdb4fabff372)

My account in Pocket-ID (The OIDC provider) uses a completely different username than than the Home Assistant admin account (My IRL first name) 

To resolve this, I created a username with my IRL name in Pocket-ID and it worked as expected in full!    

All of this to say, your README which I used may need to include a note about some systems using the required Client Secret.  I though that it was a bit weird when I set it up last night, but I wanted to try.    Additionally, the callback URL wasn't specified in the README (Unless I missed it!)  :)   Also here https://github.com/christiaangoossens/hass-oidc-auth/discussions/3

So, for Pocket-ID:

If you use a different login than your Home Assistant user, you will need to create it and set it up first. 
Set the callback in Pocket ID to Match replacing the HASSURL with your external Home Assistant URL:
![image](https://github.com/user-attachments/assets/8a7169db-e794-4c68-b2ea-46022a4e8b6f)

In the Home Assistant configuration.yaml :   (BE AWARE OF ANY YAML SPACING!)  

```
  auth_oidc:
  client_id: "YOUR CLIENT ID copied from the above screen"
  client_secret: "YOUR SUPER SECRET CLIENT Secret copied from the above screen:"
  discovery_url: "https://Pocket-ID_URL.com/.well-known/openid-configuration"
  features:
      automatic_user_linking: true
```


Then, save and restart Home Assistant and test!

**Reply by christiaangoossens** [reply]
@Node815 Some notes on your discoveries:

- You can either configure it as **confidential** (the default setting apparently for Pocket ID, in which case you indeed need the client secret) or **public** (the checkbox above the Home Assistant logo in your last screenshot), which was what the README is currently written for (as it's a bit easier to setup). So, if you would have checked that checkbox, you could have left out the client secret.
  - Also make sure to use the `!secret` helpers such that you never accidentally share your secret (it's effectively still visible in your screenshot as well)
- The issue with the groups (the NoneType error) should be resolved in the next patch release.
   - I will also make the group configurable, so if Pocket ID supports groups you can get admin on your new HA account automatically if you have that group in Pocket ID
  - If you don't have important configuration in your account, this would allow you to use your normal OIDC account instead of the "same-username" one you created.
  - You should disable automatic user linking again after you are done for better security, especially if you delete the "same-username" account in Pocket ID later on.
- I will include the callback URL in the next update of the README, thank you for the suggestion!

**Reply by Node815** [reply]
The client ID is the public ID so I didn't see the need to obscure that in my screenshot. :)  

Pocket ID does have group support, I have admin and standard users.  (It's my family) so, myself, wife and daughter who I will be migrating their mobile login to the external Home Assistant login after I get the redirect set up to that URL.    Once they are set up, I will disable the user linking. 

Thank you so much for making this plugin!   I voted a while back in the open letter for OIDC and it does seem Nabu Casa is not too motivated in making this possible so having this as an option is a welcome relief!! :+1:

**Reply by christiaangoossens** [reply]
> The client ID is the public ID so I didn't see the need to obscure that in my screenshot. :)  

Haha, no I meant that I can just about guess the secret characters because the blur is slightly too high.

> Pocket ID does have group support, I have admin and standard users.  (It's my family) so, myself, wife and daughter who I will be migrating their mobile login to the external Home Assistant login after I get the redirect set up to that URL.    Once they are set up, I will disable the user linking. 

If you can, it would be very helpful if you can enable debug mode (HA docs have details on how to do that) and send me a copy of your id token with groups in it, so I can see the format. It should be logged if debug is on. Otherwise, I will get around to that when I test pocket id myself.

> Thank you so much for making this plugin!   I voted a while back in the open letter for OIDC and it does seem Nabu Casa is not too motivated in making this possible so having this as an option is a welcome relief!! :+1:  

Thank you! Hope it helps. Make sure to wait a little bit until at least beta to migrate your family, as things might still change or break. Gotta keep that Wife Approval Factor ;)

**Reply by Node815** [reply]
> If you can, it would be very helpful if you can enable debug mode (HA docs have details on how to do that) and send me a copy of your id token with groups in it, so I can see the format. It should be logged if debug is on. Otherwise, I will get around to that when I test pocket id myself.
> 
I run the Home Assistant OS in Promox so I can log into the container and follow logs.  I did enable the logger: in the configuration.yaml though for the component and captured an error for the Pocket-ID Admin group user but don't think that would be of help here.  (it's mostly async and python errors in the aiohttp and http components. )  Which are more than likely due to the grouping issue.  

> Thank you! Hope it helps. Make sure to wait a little bit until at least beta to migrate your family, as things might still change or break. Gotta keep that Wife Approval Factor ;)

Of course!  I tend to live on the bleeding edge of things on my systems in running beta/alpha stuff periodically.  Even my OS is on the bleeding edge and a rolling release distro!   My wife only uses her mobile device for anything she does, she rarely uses the HA dashboard except to control our lights.  :)    We use it for real-time geolocation of our devices when we are out in the metro areas here, so we can know where a family member is.   Especially helpful for anxiety prone members of my family!

**Comment by begleyt** [comment]
Thanks for what you created. 

I was able to get this to work with NextCloud OIDC client add on app.  

I had to align the Display name in nextcloud with the username in home assistant to get the two exisitng accounts to link.  I utlized the claims parameters to accomplish that.

```
auth_oidc:
  client_id: "{clientID}"
  discovery_url: "https://{yourdomain}/.well-known/openid-configuration"
  display_name: "Login with NextCloud"
  features:
      automatic_user_linking: true
  claims:
    display_name: preferred_username
    username: name
```

**Comment by vilmosnagy** [comment]
I just installed my first ever home assistant instance, and I'm very happy that I can use my pre-existing keycloak instance for login (with this my wife does not have to remember passwords).

But is there a way to have something similar to Home Assistant's "Keep me logged in" feature? With the current setup I have to click the login button on every page reload, which is not a huge hassle, but still a bit annoying. Thanks,

**Reply by christiaangoossens** [reply]
I do stay logged in for page reloads/new tabs, can you give me more details on when you get signed out?

**Reply by vilmosnagy** [reply]
@christiaangoossens I'll try to upload a video here; it logs me out on a simple reload / `F5` key press. The OS is up-to-date Fedora, with an up-to-date Chrome as the browser. I've cropped the URL of my HA server, but if it helps then I try to record a video with the full screen and just the URL blurred.

I didn't have the time to debug more, but if you could point me in a few directions, I'm willing to give it a go. I'm new with Home Assistant, and I have a very small Python experience (I've developed a few Django apps ~10ish years ago as a junior dev).

[home-assistant-reload-logout.webm](https://github.com/user-attachments/assets/4fc93095-b61c-4712-8e7f-b55f67a5d1ef)

**Reply by christiaangoossens** [reply]
@vilmosnagy Can you login with a normal HA account without selecting anything special such as 'keep me logged in'? This seems like an HA bug or a cookie/proxy error.

**Reply by vilmosnagy** [reply]
@christiaangoossens hm, you seem to be right. Without selecting the `keep me logged in` function HA logs me out instantly on the first page reload - I always tick these _keep me logged in_ boxes everywhere, that's why I didn't found this issue.

Isn't that suppose to happen? Thanks for the info, gonna take a look at HA's repositories for a similar issue.

**Reply by christiaangoossens** [reply]
@vilmosnagy Logging in without that checkmark should keep you logged in for 30 minutes (which it does seem to do on my own install), so your OIDC login should stay there for 30 min. If you check 'Keep me logged in', you will stay logged in for longer (indefinitely?)

Your browser might delete your local storage on every refresh, might be a good place to start your search.

**Reply by christiaangoossens** [reply]
This might actually have been https://github.com/christiaangoossens/hass-oidc-auth/issues/70.

**Comment by Dreanaught** [comment]
I use it with authentik and have no problems.
The only thing i wonder is the discovery-url. On other services, immich for example, the "/.well-known/openod-configuration"-part is not required.

```
auth_oidc:
  client_id: "clientID"
  client_secret: "clientSecret"
  discovery_url: "https://domain.tld/application/o/home-assistant/.well-known/openid-configuration"
  features:
    automatic_user_linking: false
```

**Reply by christiaangoossens** [reply]
You're right, that part is defined, so there is no technical reason we need it. Just chose to include it because it's easy to recognise and to instruct people to copy the '.well-known/openid-configuration' URL.

**Comment by deftmartian** [comment]
Greate work on this.  I use Authentik.  Was using a buggy forward auth config previously.

It took me a minute to figure out how to fix the following error:
```
2025-02-23 15:48:59.535 WARNING (MainThread) [custom_components.auth_oidc.oidc_client] ID Token received signed with the wrong algorithm: ES256, expected RS256
2025-02-23 15:48:59.535 WARNING (MainThread) [custom_components.auth_oidc.oidc_client] Failed to complete token flow, returning None. ()
```
Had debugging enabled, but didn't realize it was a config option.  My cert is signed by letsencrypt

**Reply by christiaangoossens** [reply]
The best way to resolve this is within Authentik by enabling the signing certificate there. Still has to be added to [the guide](https://github.com/christiaangoossens/hass-oidc-auth/blob/main/docs/provider-configurations/authentik.md) too.

**Reply by deftmartian** [reply]
Sorry, I wasn't very clear.  I had enabled letsencrypt cert as the signing key in the Authentik provider.   The hass-oidc-auth integration expects a RS256 key by default.  

once I updated my config to change this to ES256, everything worked as expected.

my config:
```
auth_oidc:
    client_id: "example"
    client_secret: !secret oidc_client_secret
    discovery_url: "https://example.com/.well-known/openid-configuration"
    id_token_signing_alg: "ES256"
```

**Comment by girlpunk** [comment]
Tried setting this up the other day, and I have to say it's wonderful! The slightly more complicated login process is a little disappointing, but I understand that's entirely down to decisions made by the Home Assistant maintainers rather than anything to do with the OIDC plugin itself.

Did have two small thoughts towards improvements, the `automatic_user_linking` is nice, however my username in my IdP has a capital letter, however Home Assistant does not support this. Is it possible to link the users manually or specify a transform filter of some kind? I guess I could add an attribute to the user and expose it as a claim, but that would require a manual setup for each user, which is less than ideal.

Second thought (which likely relies on the other UI improvements going through), is it possible to disable Home Assistant's built-in authentication?

Thank you for improving yet another feature Home Assistant has stubbornly refused to implement sensibly for years.

**Reply by christiaangoossens** [reply]
Thanks for your feedback! For the user linking, please see https://github.com/christiaangoossens/hass-oidc-auth/issues/58#issuecomment-2763293136 for the reasons why an automatic or manual transform may be problematic.

Adding UI elements to link manually is also very hard, but it would be possible to add a config array to map subjects (oidc sub) to usernames in HA which you would have to configure manually before first login. Although, at that point you might as well just configure the manual attribute and claim setup in Authentik.

Generally usernames with capital letters are not a great idea on the Authentik side anyway, many apps don't support them and usernames are generally always case sensitive (different from user handles, which are not).

Adding a flag to disable normal login is possible, but I don't think its a great idea because it might lock users out. Having a local backup user if the integration malfunctions is always a good idea. Especially if you add another domain redirect to the OIDC start url, normal users will never see the option either. I don't want to encourage disabling the normal auth for most users because of those reasons.

Let me know how you feel about them.

**Reply by girlpunk** [reply]
> Having a local backup user if the integration malfunctions is always a good idea.

Oh definitely, but given the lovely username and password boxes available, having it available all the time seems like a great way to confuse end users. Having the backup account is a good idea, but one could always temporarily remove the flag if needed to re-enable local auth

**Comment by caplam** [comment]
Hello,
Thanks for your effort. Sso is really a wanted feature in a homelab.
I followed instructions but i think i missed something.
When logging from a private window on the lan with "haos.my.domain.com"
i'm presented a window with a code to enter or click to login with homeassistant local.
But no matter which code i enter login is denied. What code should it be ? i tried with my authentik account totp and with home assistant totp  (obviously this is not this one).
It seems callback uri is not correct:
In authentik i entered that one : "https://haos.mydomain.com/auth/oidc/callback" and what i have in my browser private window is
"https://haos.mydomain.com/auth/authorize?response_type=code&redirect_uri=https%3A%2F%2Fhaos.mydomain.com%2F%3Fauth_callback%3D1&client_id=https%3A%2F%2Fhaos.mydomain.com%2F&state=eyJoYXNzVXJsIjoiaHR0cHM6Ly9oYW9zLnJrMS5mciIsImNsaWVudElkIjoiaHR0cHM6Ly9oYW9zLnJrMS5mci8ifQ%3D%3D"
my config in home assistant configuration.yaml:
`auth_oidc:
  client_id: "myclientid"
  client_secret: !secret oidc_client_secret
  discovery_url: "https://auth.mydomain.com/application/o/home-assistant/.well-known/openid-configuration"
  display_name: "Authentik"
  features:
    automatic_user_linking: true
  `
When opening home assistant from authentik dashboard it works as expected.
When opening https://haos.mydoamin.com/auth/oidc/welcome it also works.
My reverse proxy is caddy 
domain is https://haos.mydomain.com
handler is http://local-ip:8123

**Reply by christiaangoossens** [reply]
Please see https://github.com/christiaangoossens/hass-oidc-auth/blob/main/docs%2Fusage.md for usage instructions.

There is no way to start the flow from the code screen as HA doesn't support custom UI for integrations on that screen. It's only used if you have already gotten the code at the end of the /auth/oidc/welcome flow, for instance if you want to login to the mobile app.

**Reply by caplam** [reply]
ok i misunderstood the part about starting the flow.
Hopefully home assistant guys will end up doing something about that.
I progressively integrate my apps (14 so far) within authentik and it's a real plus.

**Reply by christiaangoossens** [reply]
Yeah, best way to have this work is to add /auth/oidc/redirect as the path behind the icon in Authentik/another dashboard so you don't have to think about it.

The code flow is there just for watches, wall tablets and phones where you cannot use OIDC within the app.

**Comment by ratermir** [comment]
Hello
Christiaan, please accept my huge thanks for this component :). It is long term awaited and needed. 
I installed it this weekend on my testing HA instance, and in collaboration with Zitadel IDP it works like a charm.
Non-existing user has been created, existing user is loged-in and paired, everything seem to be OK.

I want to ask whether did you think about how the final screes setup will look like? By me, it would be perfect to have these options:
- option to configure HA to bypass HA login screen completely and redirect directly to IDP login screen (or log-in directly if preconditions - valid token - exist)
but in this case there must be an option to allow "direct HA" login (URL ...?) in case IDP issue
- option to configure HA login screen to choose login method (Direct HA, login with IDP - reasonably named), in this case it would be possible to allow multiple OIDC providers (but I don't need it - only thinking)

I suppose that current way is dev only (to have completly separate login URL for SSO).

**Reply by christiaangoossens** [reply]
Right now there is no progress in that direction because the HA team refuses any contributions to enable that. It's tracked in https://github.com/christiaangoossens/hass-oidc-auth/issues/19.

So, I expect the current experience with some tweaks to move on to a beta stage, as soon as we have tests. It's not the prettiest, but it works on all platforms.

**Comment by Rihan9** [comment]
Hi!
I just finished to setup this addon and it works just fine! For info, I'm using Authelia.
I liked the unexpected two-step login flow, it should prevent the companion app to freak out like the addon that enables you to use the header authentication.

Honestly, the only things that I saw that need to be fixed are:
- on the login page, next to the login button should also be present a button to redirect the user to /auth/oidc/welcome. Next step for this button should be to identify the device and, if mobile, open the link in a separate browser.
- still in the login page, "just checking" really need to be changed :D.
- on the /auth/oidc/welcome page, the button "log ito home assistant in this browser" can be paired to "log on home assistant in the companion app" if the device is a mobile's one (you should be able to check this using the headers). next step for this can be a simple renaming, because,  onestly, the button name are way too long. You can rename them in "continue in this broser" and "continue in the companion app" or something similar. 

If you are able to change these buttons in the login page and the buttons in the /auth/oidc/welcome page, the needs to use a separate device shouldn't be needed anymore and you can pass the code directly as a parameters when in the /auth/oidc/welcome the user click one of these buttons suggested. 

So much works to do! I'm sorry. Treat this comment as "bunch of ideas to stimulate other ideas" :D

**Reply by christiaangoossens** [reply]
None of these are currently possible, please see https://github.com/christiaangoossens/hass-oidc-auth/issues/19 and the issues/discussion posts that link to it for more information.

**Reply by christiaangoossens** [reply]
More info about your last idea here: https://github.com/christiaangoossens/hass-oidc-auth/issues/59#issuecomment-2764635896

**Reply by christiaangoossens** [reply]
Also thank you for writing it, I just get a lot of these 🙃, with no way to solve them without weird hacks.

**Comment by lakemike** [comment]
Hi, I tried to implement everything as per the documentation.

homeassistant's configuration.yaml
```
auth_oidc:
  client_id: "homeassistant"
  client_secret: "insecure_secret"
  discovery_url: "https://authelia.mydomain.tld/.well-known/openid-configuration"
```


authelia 4.39 configuration.yaml
```
      - client_id: 'homeassistant'
        claims_policy: 'default'
        client_name: 'Home Assistant'
        client_secret: '$pbkdf2-sha512$310000$c8p78n7pUMln0jzvd4aK4Q$JNRBzwAo0ek5qKn50cFzzvE9RXV88h1wJn5KGiHrD0YKtZaR/nCb2CJPOsKaPK0hjf.9yHxzQGZziziccp6Yng'  # The digest of 'insecure_secret'.
        public: false
        require_pkce: true
        pkce_challenge_method: 'S256'
        authorization_policy: 'two_factor'
        consent_mode: pre-configured
        redirect_uris:
          - 'https://hass.mydomain.tld/auth/oidc/callback'
        scopes:
          - 'openid'
          - 'profile'
          - 'groups'
        userinfo_signed_response_alg: 'RS256'
        token_endpoint_auth_method: 'client_secret_post'
```
I do get to the button that says "SSO". Clicking it brings up the authelia login.

But after authenticating, I get

```
500 Internal Server Error
Server got itself in trouble
```

Any idea where I got it wrong?

**Reply by christiaangoossens** [reply]
Which doc did you follow? User info shouldn't be signed, you can remove that line.

**Reply by lakemike** [reply]
oh, that did it. I must have made a mistake adding the oidc client. Many thanks! It's working now!!

**Comment by NoelGisler** [comment]
Are newly created users supposed to not have a username? Because it makes oidc-created users incompatible with browser-mod.

**Reply by christiaangoossens** [reply]
Yes, only users with a password have a username in Home Assistant. As users for OIDC only have an OIDC id (sub) that they login with, they have neither a username nor password. Looks strange in the UI, but it's fully compliant with the way the built-in auth provider works.

Feels like a strange requirement to me, can you send a link to the other integration? New users have a user id and even a person entity is created, so other integrations should just use those.

**Reply by NoelGisler** [reply]
The Link to Browser-mods HACS-Rpository? gladly: https://github.com/thomasloven/hass-browser_mod

I am not sure it's refusing to work because of the missing username, maybe I hastily jumped to conclusions. But when I try to set a setting for a user, oidc-created users don't shop up in the dropdown, and the biggest difference I saw is the missing username.

**Reply by christiaangoossens** [reply]
I created an issue over on that repository for this bug to be fixed: https://github.com/thomasloven/hass-browser_mod/issues/1102. It's not a bug here, but instead a handling error in that integration. Should be very fixable.

**Reply by dcapslock** [reply]
Browser Mod [v2.7.0-beta.3](https://github.com/thomasloven/hass-browser_mod/releases/tag/v2.7.0-beta.3) available with an update to address this issue.

**Reply by NoelGisler** [reply]
Can confirm the Beta works. That was incredibly fast, thank you 🥳
